bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/messages/ParadoxIPPacket.java

```diff
@@ -26,173 +26,128 @@
         return this;
     }
 
     public ParadoxIPPacket setMessageType(HeaderMessageType messageType) {
-        this.messageType = messageType.getValue();
+        header.messageType = messageType.getValue();
         return this;
     }
 
-    public byte getEncryption() {
-        return encryption;
-    }
-
-    public ParadoxIPPacket setEncryption(byte encryption) {
-        this.encryption = encryption;
+    public ParadoxIPPacket setUnknown0(byte unknownByteValue) {
+        header.unknown0 = unknownByteValue;
         return this;
     }
 
-    public byte getCommand() {
-        return command;
-    }
-
-    public ParadoxIPPacket setCommand(HeaderCommand command) {
-        this.command = command.getValue();
-        return this;
-    }
-
-    public ParadoxIPPacket setCommand(byte command) {
-        this.command = command;
-        return this;
+    @Override
+    public PacketHeader getHeader() {
+        return header;
     }
 
-    public byte getSubCommand() {
-        return subCommand;
+    @Override
+    public byte[] getPayload() {
+        return payload;
     }
 
-    public ParadoxIPPacket setSubCommand(byte subCommand) {
-        this.subCommand = subCommand;
-        return this;
+    @Override
+    public void encrypt() {
+        EncryptionHandler encryptionHandler = EncryptionHandler.getInstance();
+        payload = encryptionHandler.encrypt(payload);
+        header.encryption = 0x09;
     }
 
-    public byte getUnknown0() {
-        return unknown0;
+    @Override
+    public String toString() {
+        return "ParadoxIPPacket [" + ParadoxUtil.byteArrayToString(getBytes()) + "]";
     }
 
-    public ParadoxIPPacket setUnknown0(byte unknown0) {
-        this.unknown0 = unknown0;
-        return this;
-    }
+    public class PacketHeader {
 
-    public long getTheRest() {
-        return theRest;
-    }
+        public PacketHeader(short payloadLength) {
+            this.payloadLength = payloadLength;
+        }
 
-    public ParadoxIPPacket setTheRest(long theRest) {
-        this.theRest = theRest;
-        return this;
+        /**
+         * Start of header - always 0xAA
+         */
+        private byte startOfHeader = (byte) 0xAA;
+
+        /**
+         * Payload length - 2 bytes (LL HH)
+         */
+        private short payloadLength = 0;
+
+        /**
+         * "Message Type: 0x01: IP responses 0x02: Serial/pass through cmd response
+         * 0x03: IP requests 0x04: Serial/pass through cmd requests"
+         */
+        private byte messageType = 0x03;
+
+        /**
+         * "IP Encryption 0x08: Disabled 0x09: Enabled"
+         */
+        private byte encryption = 0x08;
+        private byte command = 0;
+        private byte subCommand = 0;
+        private byte unknown0 = 0x00;
+        private byte unknown1 = 0x01;
+
+        public byte[] getBytes() {
+            try {
+                ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```

Comment:
Why not just use a ByteBuffer here instead? You are already using one to append the payloadLength. Another option is to wrap your ByteArrayOutputStream in a DataOutputStream and use that to append short values and the like.

Reply:
Hi, as I never used bytebuffer before I've just checked when it's used and some examples.
Is this example that I've created correct usage of the ByteBuffer?
`  ByteBuffer buf = ByteBuffer.wrap(header.getBytes());
        buf.put(payload);
        return buf.array();`
The code becomes way simpler and I like it much more but I'm not sure if it's OK to return the backing array or I should create a new one here (clone of the backing array). I guess it's fine to return it as it's used only here...

Reply:
Just saw another example in stackoverflow which looks more correct, i.e. this one. To me this should be the right way as probably the backing array may be longer than the actual bytes in the buffer... What do you think? `ByteBuffer buf = ...
byte[] arr = new byte[buf.remaining()];
buf.get(arr);`

Reply:
The safest route would be:
```
byte[] bufferArray = new byte[BUFFERLENGTH];
ByteButter buf = ByteBuffer.wrap(bufferArray);
...
buf.put(...);
buf.put(...);
...
return bufferArray;
```
You just need to make sure that you know the exact size of byte array you need ahead of time.